nsxt/resource_nsxt_upgrade_prepare.go

```diff
@@ -0,0 +1,378 @@
+							Type:        schema.TypeString,
+							Description: "Message of the failed precheck",
+							Computed:    true,
+						},
+					},
+				},
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceNsxtUpgradePrepareCreate(d *schema.ResourceData, m interface{}) error {
+	id := d.Id()
+	if id == "" {
+		id = newUUID()
+	}
+	err := prepareForUpgrade(d, m)
+	if err != nil {
+		return handleCreateError("NsxtUpgradePrepare", id, err)
+	}
+	d.SetId(id)
+	return resourceNsxtUpgradePrepareRead(d, m)
+}
+
+func prepareForUpgrade(d *schema.ResourceData, m interface{}) error {
+	// 1. Upload upgrade bundle and wait for upload to complete
+	err := uploadPrecheckAndUpgradeBundle(d, m)
+	if err != nil {
+		return logAPIError("Failed to upload bundle", err)
+	}
+	// 2. Accept eula
+	err = acceptUserAgreement(d, m)
+	if err != nil {
+		return err
+	}
+	// 3. Upgrade UC and check for its upgrade status
+	err = upgradeUc(d, m)
+	if err != nil {
+		return logAPIError("Failed to upgrade Upgrade Coordinator", err)
+	}
+	// 4. Execute pre-upgrade checks
+	err = executePreupgradeChecks(d, m)
+	if err != nil {
+		return logAPIError("Failed to execute pre-upgrade checks", err)
+	}
+	// 5. Check if there is error in prechecks
+	precheckFailures, err := getPrecheckErrors(m, nsxModel.UpgradeCheckFailure_TYPE_FAILURE)
+	if err != nil {
+		return logAPIError("Failed to get precheck errors", err)
+	}
+	if len(precheckFailures) > 0 {
+		var failureIDs []string
+		for _, failure := range precheckFailures {
+			failureID := *failure.Id
+			failureIDs = append(failureIDs, failureID)
+		}
+		return fmt.Errorf("Encounter errors in precheck %s, please address these error on NSX UI", failureIDs)
+	}
+	return nil
+}
+
+func resourceNsxtUpgradePrepareRead(d *schema.ResourceData, m interface{}) error {
+	id := d.Id()
+	precheckWarnings, err := getPrecheckErrors(m, nsxModel.UpgradeCheckFailure_TYPE_WARNING)
+	if err != nil {
+		return handleReadError(d, "NsxtUpgradePrepare", id, err)
+	}
+	err = setFailedPrechecksInSchema(d, precheckWarnings)
+	if err != nil {
+		return handleReadError(d, "NsxtUpgradePrepare", id, err)
+	}
+	return nil
+}
+
+func resourceNsxtUpgradePrepareUpdate(d *schema.ResourceData, m interface{}) error {
+	id := d.Id()
+	err := prepareForUpgrade(d, m)
+	if err != nil {
+		return handleUpdateError("NsxtUpgradePrepare", id, err)
+	}
+	return resourceNsxtUpgradePrepareRead(d, m)
+}
+
+func resourceNsxtUpgradePrepareDelete(d *schema.ResourceData, m interface{}) error {
+	return nil
+}
+
+func uploadPrecheckAndUpgradeBundle(d *schema.ResourceData, m interface{}) error {
+	upgradeBundleURL := d.Get("upgrade_bundle_url").(string)
+	precheckBundleURL := d.Get("precheck_bundle_url").(string)
+	upgradeBundleType := nsxModel.UpgradeBundleFetchRequest_BUNDLE_TYPE_UPGRADE
+	precheckBundleType := nsxModel.UpgradeBundleFetchRequest_BUNDLE_TYPE_PRE_UPGRADE
+	version := d.Get("version").(string)
+	c := m.(nsxtClients)
+	userName := c.NsxtClientConfig.UserName
```

Comment:
other types of auth are not supported?
We'll need to clarify the docs about limitations of upgrade config.

Reply:
I'm afraid that's the case, will add this to the doc.
Since nsxt-mp/v0.4.0, password, username, version and bundle type are added to the bundleFetchRequest schema, before v0.4.0 only URL is required. (I think nsxt-mp/v0.4.0 corresponds to nsx version 4.1.1 or higher because that's when NSX starts to support uploading pub type bundle)